File: Algo_DBMS/Algo_Lab_Sheet_30/lab_Sheet_Three.cpp

# include <iostream>
# include <vector>
# include <algorithm>
# include <cstdlib>
# include <ctime>
using namespace std;

// Recursive Binary Search Function
int recursiveBinarySearch (const vector <int> & array, int low, int high, int target)
{
    if (low > high)
    {
        return -1; // Element not found
    }

    int mid = low + (high - low) / 2; // Avoids potential overflow
    if (array [mid] == target)
    {
        return mid; // Element found at mid
    }
    else if (array [mid] > target)
    {
        return recursiveBinarySearch (array, low, (mid - 1), target); // Search left half
    }
    else
    {
        return recursiveBinarySearch (array, (mid + 1), high, target); // Search right half
    }
}

// Measure time taken for different input sizes
void measureTime ()
{
    vector <int> inputSizes = {100, 500, 1000, 5000, 10000, 50000, 100000};

    for (int size : inputSizes)
    {
        // Generate a sorted list of random numbers
        vector <int> array (size);
        srand (time(0)); // Seed for random numbers
        for (int i = 0; i < size; i++)
        {
            array [i] = rand() % 100000 + 1; // Random values between 1 and 100000
        }
        sort (array.begin(), array.end()); // Sort the array

        // Choose a random target element
        int target = array [rand() % size];

        // Measure execution time
        clock_t start_time = clock ();
        int result = recursiveBinarySearch (array, 0, (size - 1), target);
        clock_t end_time = clock ();
        double elapsed_time = double (end_time - start_time) / CLOCKS_PER_SEC;

        cout << "Input Size :  " << size
             << ", Target :  " << target
             << ", Index :  " << result
             << ", Time Taken :  " << elapsed_time << " seconds" << endl;
    }
}

// Main function
int main()
{
    measureTime();
    return 0;
}

File: Algo_DBMS/Algo_Lab_Sheet_30/lab_Sheet_Three.py

import time
import random
import matplotlib.pyplot as plt

# Recursive Binary Search Function
def recursive_binary_search (array, low, high, target) :
    if (low > high) :
        return -1  # Element not found
    
    mid = ((low + high) // 2)
    if (array [mid] == target) :
        return mid  # Element found at mid
    elif (array [mid] > target) :
        return recursive_binary_search (array, low, (mid - 1), target)  # Search left half
    else :
        return recursive_binary_search (array, (mid + 1), high, target)  # Search right half

# Measure time taken for different input sizes
def measure_time () :
    input_sizes = [100, 500, 1000, 5000, 10000, 50000, 100000]
    times = []
    
    for size in input_sizes :
        # Generate a sorted list of random numbers
        array = sorted ([random.randint (1, 100000) for _ in range (size)])
        
        # Choose a random target element
        target = random.choice (array)
        
        # Measure execution time
        start_time = time.time ()
        result = recursive_binary_search (array, 0, (len (array) - 1), target)
        end_time = time.time ()
        elapsed_time = (end_time - start_time)
        
        print (f"Input Size :  {size}, Target :  {target}, Index :  {result}, Time Taken :  {elapsed_time:.6f} seconds")
        times.append (elapsed_time)
    
    # Plot graph
    plt.plot (input_sizes, times, marker='o')
    plt.xlabel ("Input Size")
    plt.ylabel ("Time Taken (seconds)")
    plt.title ("Recursive Binary Search: Input Size vs Time Taken")
    plt.grid (True)
    plt.show ()

# Run the experiment
if __name__ == "__main__" :
    measure_time ()
